Archive/Imo/Imo2025Q3.lean

```diff
@@ -0,0 +1,255 @@
+  have : IsUnit (-1 : ZMod k) := by simp
+  obtain ⟨p, hp⟩ := Nat.forall_exists_prime_gt_and_eq_mod this n
+  use p
+  refine ⟨hp.2.1, hp.1, ?_⟩
+  rw [modEq_comm, ← ZMod.intCast_eq_intCast_iff, ZMod.intCast_eq_intCast_iff_dvd_sub, Int.sub_neg]
+  have := hp.2.2
+  rw [eq_neg_iff_add_eq_zero, ← cast_add_one p, ZMod.natCast_eq_zero_iff] at this
+  exact ofNat_dvd_left.mpr this
+
+theorem bonza_apply_prime_gt_two_eq_one (hf : f ∈ bonza) (hnf : ¬ ∀ x, x > 0 → f x = x) :
+    ∀ p > 2, Nat.Prime p → f p = 1 := by
+  obtain ⟨N, hN⟩ : ∃ N, ∀ p > N, Nat.Prime p → f p = 1 :=
+    bonza_not_x_apply_prime_of_gt_eq_one hf hnf
+  have apply_dvd_pow_sub {a p: ℕ} (ha : a > 0) (pp : Nat.Prime p) (hp : p > N) :
+      (f a : ℤ) ∣ p ^ a - 1 := by
+    have := hf.1 a p ha (by omega)
+    rwa [hN p hp pp, Nat.cast_one, one_pow] at this
+  intro q hq qp
+  obtain ⟨k, ha1, ha2⟩ : ∃ k, k ≤ q ∧ f q = q ^ k :=
+    (Nat.dvd_prime_pow qp).mp (bonza_apply_dvd_pow hf (zero_lt_of_lt hq))
+  by_cases ch : k = 0
+  · simpa [ch] using ha2
+  · have {p : ℕ} (pp : Nat.Prime p) (hp : p > N) : (q : ℤ) ∣ p ^ q - 1 := by calc
+      _ ∣ (f q : ℤ) := by
+        rw [ha2, Int.natCast_pow q k]
+        exact dvd_pow_self (q : ℤ) ch
+      _ ∣ _ := apply_dvd_pow_sub (zero_lt_of_lt hq) pp hp
+    obtain ⟨p, hp⟩ := Nat.exists_prime_gt_modEq_neg_one N (NeZero.of_gt hq)
+    have : 1 ≡ -1 [ZMOD q] := by calc
+      _ ≡ p ^ q [ZMOD q] := by
+        rw [Int.modEq_iff_dvd]
+        exact this hp.1 hp.2.1
+      _ ≡ p [ZMOD q] := ModEq.pow_card_eq_self qp
+      _ ≡ _ [ZMOD q] := hp.2.2
+    rw [modEq_comm, Int.modEq_iff_dvd] at this
+    have : (q : ℤ).natAbs ≤ (1 - (-1) : ℤ).natAbs := natAbs_le_of_dvd_ne_zero this (by norm_num)
+    omega
+
+lemma bonza_not_id_two_pow (hf : f ∈ bonza) (hnf : ¬ ∀ x, x > 0 → f x = x) :
+    ∀ n, n > 0 → ∃ a, f n = 2 ^ a := fun n hn ↦ by
+  have : ∀ {p}, Nat.Prime p → p ∣ f n → p = 2 := by
+    intro p pp hp
+    by_contra nh
+    have dvd : (p : ℤ) ∣ p ^ n - 1 := by calc
+      _ ∣ (f n : ℤ) := ofNat_dvd.mpr hp
+      _ ∣ _ := by
+        have := hf.1 n p hn (Prime.pos pp)
+        have p_gt_two : p > 2 := Nat.lt_of_le_of_ne (Prime.two_le pp) (fun a ↦ nh (id (Eq.symm a)))
+        rwa [bonza_apply_prime_gt_two_eq_one hf hnf p p_gt_two pp, Nat.cast_one, one_pow] at this
+    have : (p : ℤ) ∣ p ^ n := dvd_pow (Int.dvd_refl p) (Nat.ne_zero_of_lt hn)
+    exact (Nat.Prime.not_dvd_one pp) (ofNat_dvd.mp ((Int.dvd_iff_dvd_of_dvd_sub dvd).mp this))
+  use (f n).primeFactorsList.length
+  exact Nat.eq_prime_pow_of_unique_prime_dvd (Nat.ne_zero_of_lt (hf.2 n hn)) this
+
+/-- An example of a bonza function achieving the maximum number of values of `c`.
+-/
+def fExample : ℕ → ℕ := fun x ↦
+  if ¬ 2 ∣ x then 1
+  else if x = 2 then 4
+  else 2 ^ (padicValNat 2 x + 2)
+
+lemma LTE_lemma_of_pow_sub {a b : ℕ} (h1b : 1 < b) (hb : ¬2 ∣ b) (ha : a ≠ 0) (Evena : Even a) :
+    (padicValNat 2 a + 2) ≤ padicValNat 2 (b ^ a - 1) := by
+  have dvd : 2 ∣ b - 1 := by
+    simp at hb
```

Comment:
```suggestion
    simp only [Nat.two_dvd_ne_zero] at hb
```
Non-terminal simp should be squeezed (not sure if the same rule is applied for Archive. Please correct me if I am wrong)